### ISSUE TITLE
torch.compile(create_block_mask) has pathological cold compile time for FlexAttention causal block masks at some sequence lengths

### ISSUE THREAD

**Issue by mirceamironenco** [pytorch/pytorch#182060]
### 🐛 Describe the bug

`torch.compile(torch.nn.attention.flex_attention.create_block_mask)` can take several
  minutes or appear to hang for a simple causal mask at some sequence lengths.

  This reproduces with a minimal script using only `create_block_mask` and a trivial
  causal `mask_mod`.

Reproduction:

```python
import argparse
import time

import torch
from torch.nn.attention.flex_attention import create_block_mask


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("seq_len", type=int)
    return parser.parse_args()


def causal_mask_mod(
    batch: torch.Tensor,
    head: torch.Tensor,
    query_index: torch.Tensor,
    kv_index: torch.Tensor,
) -> torch.Tensor:
    return query_index >= kv_index


args = parse_args()

print("starting compile")
compiled = torch.compile(create_block_mask)


print(f"calling compiled create_block_mask with Q_LEN=KV_LEN={args.seq_len}")
call_start = time.perf_counter()
block_mask = compiled(
    causal_mask_mod,
    B=1,
    H=1,
    Q_LEN=args.seq_len,
    KV_LEN=args.seq_len,
    device="cuda",
    BLOCK_SIZE=(128, 128),
)
call_elapsed = time.perf_counter() - call_start
print(f"compiled create_block_mask returned in {call_elapsed:.6f}s")

torch.cuda.synchronize()
print("done")
```

Running this on an A100 SXM4:
```bash
rm -rf /tmp/torchinductor_root && rm -rf /root/.triton/cache
python block_mask_debug.py 13567
```

we get:

```py
starting compile
calling compiled create_block_mask with Q_LEN=KV_LEN=13567
compiled create_block_mask returned in 573.226887s
done
```
Same results for 13568, 13696. For 17407 it completes in 20s. Same results with `fullgraph=True`.

### Versions

Collecting environment information...
PyTorch version: 2.13.0.dev20260429+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: Could not collect
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.12.13 | packaged by conda-forge | (main, Mar  5 2026, 16:50:00) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-161-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 13.0.88
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA A100-SXM4-40GB
Nvidia driver version: 580.95.05
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.14.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.14.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.14.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.14.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.14.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.14.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.14.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.14.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           48 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  128
On-line CPU(s) list:                     0-127
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 7532 32-Core Processor
CPU family:                              23
Model:                                   49
Thread(s) per core:                      2
Core(s) per socket:                      32
Socket(s):                               2
Stepping:                                0
BogoMIPS:                                4790.91
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca ibpb_exit_to_user
Virtualization:                          AMD-V
L1d cache:                               2 MiB (64 instances)
L1i cache:                               2 MiB (64 instances)
L2 cache:                                32 MiB (64 instances)
L3 cache:                                512 MiB (32 instances)
NUMA node(s):                            8
NUMA node0 CPU(s):                       0-7,64-71
NUMA node1 CPU(s):                       8-15,72-79
NUMA node2 CPU(s):                       16-23,80-87
NUMA node3 CPU(s):                       24-31,88-95
NUMA node4 CPU(s):                       32-39,96-103
NUMA node5 CPU(s):                       40-47,104-111
NUMA node6 CPU(s):                       48-55,112-119
NUMA node7 CPU(s):                       56-63,120-127
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Mitigation; untrained return thunk; SMT enabled with STIBP protection
Vulnerability Spec rstack overflow:      Mitigation; safe RET
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Retpolines; IBPB conditional; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] numpy==2.4.3
[pip3] nvidia-cublas==13.1.0.3
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvtx==13.0.85
[pip3] torch==2.13.0.dev20260429+cu130
[pip3] torchaudio==2.11.0+cu130
[pip3] torchcodec==0.11.0+cu130
[pip3] torchvision==0.27.0.dev20260430+cu130
[pip3] triton==3.7.0+git88b227e2
[conda] numpy                      2.4.3                     pypi_0                pypi
[conda] nvidia-cublas              13.1.0.3                  pypi_0                pypi
[conda] nvidia-cuda-cupti          13.0.85                   pypi_0                pypi
[conda] nvidia-cuda-nvrtc          13.0.88                   pypi_0                pypi
[conda] nvidia-cuda-runtime        13.0.96                   pypi_0                pypi
[conda] nvidia-cudnn-cu13          9.20.0.48                 pypi_0                pypi
[conda] nvidia-cufft               12.0.0.61                 pypi_0                pypi
[conda] nvidia-curand              10.4.0.35                 pypi_0                pypi
[conda] nvidia-cusolver            12.0.4.66                 pypi_0                pypi
[conda] nvidia-cusparse            12.6.3.3                  pypi_0                pypi
[conda] nvidia-cusparselt-cu13     0.8.1                     pypi_0                pypi
[conda] nvidia-nccl-cu13           2.29.7                    pypi_0                pypi
[conda] nvidia-nvjitlink           13.0.88                   pypi_0                pypi
[conda] nvidia-nvtx                13.0.85                   pypi_0                pypi
[conda] torch                      2.13.0.dev20260429+cu130  pypi_0                pypi
[conda] torchaudio                 2.11.0+cu130              pypi_0                pypi
[conda] torchcodec                 0.11.0+cu130              pypi_0                pypi
[conda] torchvision                0.27.0.dev20260430+cu130  pypi_0                pypi
[conda] triton                     3.7.0+git88b227e2         pypi_0                pypi

cc @jerryzh168 @chauhang @penguinwu @oulgen @jamesjwu @aorenste @anijain2305 @laithsakka @masnesral @coconutruben @aditvenk @Chillee @drisspg @yanboliang @BoyuanFeng @liangel-02 @howardzhang-cv

**Comment by drisspg** [comment]
The extremly long slowdown appears to be triton compilation, and here is an abstracted example that can be used to reproduce. Note that this seems to be related to the huge rblock value

```python
import argparse
import os
import shutil
import tempfile
import time

import triton
import triton.language as tl
from triton.compiler import ASTSource, compile as triton_compile


@triton.jit
def causal_block_sum(out_ptr0, xnumel, r0_numel, XBLOCK: tl.constexpr, R0_BLOCK: tl.constexpr):
    xnumel = 4096
    r0_numel = 16384
    rnumel = r0_numel
    RBLOCK: tl.constexpr = R0_BLOCK
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:, None]
    xmask = tl.full([XBLOCK], True, tl.int1)[:, None]
    r0_base = tl.arange(0, R0_BLOCK)[None, :]
    rbase = r0_base
    x1 = xindex // 64
    x0 = xindex % 64
    acc = tl.full([XBLOCK, R0_BLOCK], 0, tl.int64)
    x4 = xindex
    for r0_offset in tl.range(0, r0_numel, R0_BLOCK):
        r0_index = r0_offset + r0_base
        r0_mask = r0_index < r0_numel
        roffset = r0_offset
        rindex = r0_index
        r0_3 = r0_index // 128
        r0_2 = r0_index % 128
        tmp0 = r0_3 + 128 * x1
        tmp1 = r0_2 + 128 * x0
        tmp2 = tmp0 >= tmp1
        tmp3 = tmp2.to(tl.int64)
        tmp4 = tl.broadcast_to(tmp3, [XBLOCK, R0_BLOCK])
        tmp5 = acc + tmp4
        acc = tl.where(r0_mask, tmp5, acc)
    tmp6 = tl.sum(acc, 1)[:, None]
    tl.store(out_ptr0 + x4, tmp6, None)


def compile_once(xblock: int, rblock: int) -> float:
    start = time.perf_counter()
    triton_compile(
        ASTSource(
            fn=causal_block_sum,
            signature={"out_ptr0": "*i64", "xnumel": "i32", "r0_numel": "i32"},
            constexprs={"XBLOCK": xblock, "R0_BLOCK": rblock},
        ),
        target=None,
    )
    return time.perf_counter() - start


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--xblock", type=int, default=1)
    parser.add_argument("--rblock", type=int, default=16384)
    parser.add_argument("--keep-cache", action="store_true")
    args = parser.parse_args()

    cache_dir = tempfile.mkdtemp(prefix="triton_fresh_cache_")
    os.environ["TRITON_CACHE_DIR"] = cache_dir
    print(f"TRITON_CACHE_DIR={cache_dir}", flush=True)
    print(f"Compiling XBLOCK={args.xblock} R0_BLOCK={args.rblock}", flush=True)
    try:
        print(f"compile_s={compile_once(args.xblock, args.rblock):.6f}", flush=True)
    finally:
        if args.keep_cache:
            print(f"Kept cache: {cache_dir}", flush=True)
        else:
            shutil.rmtree(cache_dir, ignore_errors=True)


if __name__ == "__main__":
    main()

```

**Comment by mirceamironenco** [comment]
IDK if this helps but when I debugged this at the time, the bottleneck for compilation turned out to be argsort, which is repeatedly called (e.g. as part of dense_to_ordered). See:

```py
import argparse
import time

import torch


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("rows", type=int, nargs="?", default=106)
    parser.add_argument("cols", type=int, nargs="?", default=106)
    parser.add_argument("--stable", action=argparse.BooleanOptionalAction, default=True)
    parser.add_argument("--dtype", choices=("int8", "int32", "bool"), default="int32")
    return parser.parse_args()


def argsort_descending_stable(x: torch.Tensor) -> torch.Tensor:
    return torch.argsort(x, dim=-1, descending=True, stable=True)


def argsort_descending_unstable(x: torch.Tensor) -> torch.Tensor:
    return torch.argsort(x, dim=-1, descending=True, stable=False)


args = parse_args()

if args.dtype == "int8":
    dtype = torch.int8
elif args.dtype == "int32":
    dtype = torch.int32
else:
    dtype = torch.bool

x = torch.randint(
    0,
    2,
    (1, 1, args.rows, args.cols),
    device="cuda",
    dtype=dtype,
)

fn = argsort_descending_stable if args.stable else argsort_descending_unstable
compiled = torch.compile(fn)

print(
    "calling compiled argsort: "
    f"shape={tuple(x.shape)}, dtype={x.dtype}, stable={args.stable}"
)
call_start = time.perf_counter()
out = compiled(x)
call_elapsed = time.perf_counter() - call_start
print(f"compiled argsort returned in {call_elapsed:.6f}s")

torch.cuda.synchronize()
print("done", out.shape, out.dtype)
```

```
$ python argsort_debug.py 106 106
calling compiled argsort: shape=(1, 1, 106, 106), dtype=torch.int32, stable=True
compiled argsort returned in 60.350678s
done torch.Size([1, 1, 106, 106]) torch.int64
```

So looking at the lowering of sort, IIUC I can force it to use the ATen CUDA sort fallback instead of the Triton variant via `TORCHINDUCTOR_PERSISTENT_REDUCTIONS=0` (cold compile):

```
$ TORCHINDUCTOR_PERSISTENT_REDUCTIONS=0 python argsort_debug.py 106 106
calling compiled argsort: shape=(1, 1, 106, 106), dtype=torch.int32, stable=True
compiled argsort returned in 1.614661s
done torch.Size([1, 1, 106, 106]) torch.int64
``` 
```
$ TORCHINDUCTOR_PERSISTENT_REDUCTIONS=0 python block_mask_debug.py 13567
starting compile
calling compiled create_block_mask with Q_LEN=KV_LEN=13567
compiled create_block_mask returned in 3.229977s
done
```
